### ISSUE TITLE
Vicious circle in pipeline_mapping.py

### ISSUE THREAD

**Issue by jscaber** [CGATOxford/CGATPipelines#86]
Dear All,
on specifying a regex for merging BAM files, ruffus returns an error (vicious circle in mergeBAMfiles).
I have narrowed this down to the following line:

``` python
MAPPINGTARGETS.append(mergeBAMFiles)
```

This is problematic as MAPPINGTARGETS is the input to mergeBAMFiles (although I don't know why this line didn't cause trouble in the past).

In my own branch, I have replaced this line with:

``` python
MAPPINGTARGETS = mergeBAMFiles
```

This however gets rid of all the picard statistics etc for the unmerged files, which may be useful to QC of the various runs per replicate.

How should we proceed in the master branch - keep both outputs or drop the unmerged?

Best wishes,
Jakub

**Comment by TomSmithCGAT** [comment]
This wasn't an issue last time I specified a regex for merging. I've looked in the ruffus release notes and I can see any relevant changes. Can you make the ruffus pipeline plot just to check for certain where ruffus thinks the vicious circle is?

I've discussed just running picard etc on the merged with Andreas before and he was of the opinion that's it's helpful to compare the seperate unmerged bams, for example if they've come from different lanes/sequencing runs.

**Comment by jscaber** [comment]
Voila.
![ruffus40](https://cloud.githubusercontent.com/assets/12160690/9689714/cdd0de00-532f-11e5-8791-99e0ba5bd511.png)

The code also worked for me up to a month ago and that particular line has been unchanged since 2013. I am not sure which recent change resulted in this error, there may well be multiple solutions...

I also attach my regex from pipeline.ini

``` python
pattern_input=(.*)(R\d)-*(\d+)*
pattern_output=\1\2
```

**Comment by TomSmithCGAT** [comment]
Is it  possible your regex leads to merged files with the same name as the unmerged infile? Could you give me an example of the unmerged and merged filenames for a sample

**Comment by jscaber** [comment]
Thanks.
Regex seems to be working as intended. I ran the identical pipeline.ini and same files a month ago on this pipeline and they worked then.
Example below:

```
Task = 'mergeBAMFiles'
    "merge BAM files from the same experiment."
       Job  = [[star.dir/M3-12-R1-1.star.bam, star.dir/M3-12-R1-2.star.bam,
               star.dir/M3-12-R1-3.star.bam, star.dir/M3-12-R1-4.star.bam,
               star.dir/M3-12-R1-5.star.bam]
             -> star.dir/M3-12-R1.star.bam]
```

**Comment by TomSmithCGAT** [comment]
Could you try it with the following:
pattern_input=(.*)-(\d)
pattern_output=\1

**Comment by jscaber** [comment]
Same error: "Circular dependencies..."
PS: It's not to do with loadSTARStats btw - if I remove the relevant reference to this, the circular dependency on mergeBAMFiles persists...

**Comment by TomSmithCGAT** [comment]
Shame, I thought the regex might be to blame. I've gotta head off. If Andreas doesn't sort it in the blink of an eye, I'll have another fumble around next week. I guess for now you can use your branch until we work out what's going on.

**Comment by IanSudbery** [comment]
I think the problem is that the post merge names also match the merge
pattern.

Sent from my iPad

On 4 Sep 2015, at 18:09, Jakub Scaber notifications@github.com wrote:

Voila.
[image: ruffus40]
https://cloud.githubusercontent.com/assets/12160690/9689714/cdd0de00-532f-11e5-8791-99e0ba5bd511.png

The code also worked for me up to a month ago and that particular line has
been unchanged since 2013. I am not sure which recent change resulted in
this error, there may well be multiple solutions...

I also attach my regex from pipeline.ini

pattern_input=(._)(R\d)-_(\d+)*pattern_output=\1\2

**Comment by TomSmithCGAT** [comment]
Yeah that's what I thought. Looks like I messed up the regex too though. These should work...

pattern_input=(.*)-(\d+)-(R\d)-(\d)
pattern_output=\1-\2-\3

**Comment by jscaber** [comment]
Thanks. No difference in the error message. The regex is similar my original regex, but more explicit. It may be difficult to imagine an output name that would not match the merge pattern partially?
I will keep looking for solutions...

**Comment by jscaber** [comment]
I played around with trying to untangle MAPPINGTARGETS.
Turns out the following works without creating a circular dependency:

```
MAPPINGTARGETS = MAPPINGTARGETS + [mergeBAMFiles]
```

A very uneducated guess: I woder if `MAPPINGTARGETS.append(mergeBAMFiles)` forces the function itself into the list. Could this be a change in syntax processing by a new version of ruffus/python?

In any case, the new code does what it's supposed to do: all subsequent functions generate a stats output for the merged files as well as the original files.

I will create a pull request.
